src/net.cpp

```diff
@@ -897,37 +897,39 @@ size_t CConnman::SocketSendData(CNode& node) const
     auto it = node.vSendMsg.begin();
     size_t nSentSize = 0;
 
-    while (it != node.vSendMsg.end()) {
-        const auto& data = *it;
-        assert(data.size() > node.nSendOffset);
+    while (true) {
+        bool progress = false;
+        if (it != node.vSendMsg.end() && node.m_transport->DoneSendingMessage()) {
+            // If possible, move one message from the send queue to the transport.
+            node.m_send_memusage -= it->GetMemoryUsage();
+            node.m_transport->SetMessageToSend(std::move(*it));
+            ++it;
+            progress = true;
+        }
+        const auto& [data, more, msg_type] = node.m_transport->GetBytesToSend();
         int nBytes = 0;
-        {
+        if (!data.empty()) {
             LOCK(node.m_sock_mutex);
             if (!node.m_sock) {
```

Comment:
I guess it's not entirely clear to me whose responsibility it is to ensure the handshake has been done: https://github.com/bitcoin/bitcoin/pull/28165#discussion_r1283674053 (I assume it will be in the main PR)

Reply:
I think `!node.m_sock` can only happen when when we've disconnected in another thread, using `CloseSocketDisconnect()` (and during tests). That's also what the documentation of `m_sock_mutex` says. cc @vasild

Reply:
8d1a91f6394f00b880613268a9f2f3164ef914e2: technically `have_unsent_data` should be set to `true` here, but since we just disconnected it doesn't matter.

Reply:
Right, if `CNode::m_sock` is empty `shared_ptr`, that means `CNode::CloseSocketDisconnect()` has been called and this is a one-way street, it is not coming back, so all communication structs and data for this peer can be dropped.

Or, during tests an empty `shared_ptr` may be supplied to the `CNode` constructor. My understanding is that in those tests the code that uses the socket is not supposed to be reached/executed. Other tests provide a mocked socket, that returns either fuzzed data or static/hardcoded contents upon read in order to entertain the code that receives from the socket.